src/raftstore/store/fsm/apply.rs

```diff
@@ -2029,95 +2013,357 @@ pub struct ApplyRes {
+            }
+
+            let entries =
+                self.delegate
+                    .load_entries_for_merge(ctx, &catch_up_logs.merge, apply_index);
+            if entries.is_empty() {
+                break 'check;
+            }
+            match self.delegate.handle_raft_committed_entries(ctx, entries) {
                 None => {
-                    error!("[region {}] is missing", apply.region_id);
-                    continue;
+                    ctx.apply_res.last_mut().unwrap().merged = true;
                 }
-                Some(e) => e.take().unwrap(),
-            };
-            delegate.metrics = ApplyMetrics::default();
-            delegate.term = apply.term;
+                Some(mut state) => {
+                    // Free redundant memory.
+                    catch_up_logs.merge.entries.clear();
+                    state.catch_up_logs = Some(catch_up_logs);
+                    self.delegate.wait_merge_state = Some(state);
+                    return;
+                }
+            }
+        }
 
-            {
-                let mut ctx = ApplyContext::new(&mut core, &mut self.delegates);
-                delegate.handle_raft_committed_entries(&mut ctx, apply.entries);
+        ctx.write_to_db();
+        catch_up_logs.ready_to_merge.store(true, Ordering::SeqCst);
+        info!("{} logs are all applied now.", self.delegate.tag);
+        let _ = catch_up_logs
+            .target_mailbox
+            .force_send(Msg::LogsUpToDate(self.delegate.region_id()));
+    }
+
+    fn handle_tasks(&mut self, apply_ctx: &mut ApplyContext, msgs: &mut Vec<Msg>) {
+        let mut channel_timer = None;
+        let mut drainer = msgs.drain(..);
+        loop {
+            match drainer.next() {
+                Some(Msg::Apply { start, apply }) => {
+                    if channel_timer.is_none() {
+                        channel_timer = Some(start);
+                    }
+                    match self.handle_apply(apply_ctx, apply) {
+                        None => {}
+                        Some(mut state) => {
+                            state.pending_msgs = drainer.collect();
+                            self.delegate.wait_merge_state = Some(state);
+                            break;
+                        }
+                    }
+                }
+                Some(Msg::Proposals(props)) => self.handle_proposals(props),
+                Some(Msg::Registration(reg)) => self.handle_registration(reg),
+                Some(Msg::Destroy(d)) => self.handle_destroy(apply_ctx, d),
+                Some(Msg::CatchUpLogs(cul)) => self.catch_up_logs_for_merge(apply_ctx, cul),
+                Some(Msg::LogsUpToDate(_)) => {}
+                #[cfg(test)]
+                Some(Msg::Validate(_, f)) => f.call_box((&self.delegate,)),
+                None => break,
             }
+        }
+        if let Some(timer) = channel_timer {
+            let elapsed = duration_to_sec(timer.elapsed());
```

Comment:
Should stop record the time before `handle_apply`.

Reply:
It's not efficient to do so. As either we are only recording the wait time of first apply command, or records wait time of every apply command. Generally, `handle_apply` is unlikely to trigger IO operation, hence record it at the end works for both performance and accuracy.